logfire/_internal/exporters/otlp.py

```diff
@@ -20,23 +20,30 @@
 
 import logfire
 
-from ..stack_info import STACK_INFO_KEYS
-from ..utils import logger, platform_is_emscripten, truncate_string
+from ..utils import logger, platform_is_emscripten
 from .wrapper import WrapperLogExporter, WrapperSpanExporter
 
 
-class OTLPExporterHttpSession(Session):
-    """A requests.Session subclass that raises a BodyTooLargeError if the request body is too large.
+class BodySizeCheckingOTLPSpanExporter(OTLPSpanExporter):
+    # 5MB is significantly less than what our backend currently accepts,
+    # but smaller requests are faster and more reliable.
+    # This won't prevent bigger spans/payloads from being exported,
+    # it just tries to make each request smaller.
+    # This also helps in case the backend limit is reduced in the future.
+    max_body_size = 5 * 1024 * 1024
 
-    Also defers failed requests to a DiskRetryer.
-    """
+    def _serialize_spans(self, spans: Sequence[ReadableSpan]) -> bytes:
+        result = super()._serialize_spans(spans)  # type: ignore
+        if len(spans) > 1 and len(result) > self.max_body_size:
+            # Tell outer RetryFewerSpansSpanExporter to split in half
+            raise BodyTooLargeError(len(result), self.max_body_size)
+        return result
```

Comment:
My only piece of feedback is that this could be expensive and that instead of re-serializing we could transform to protobuf and then manipulate the in-memory representation of the protobuf since that's cheap to manipulate and cheap to calculate the size of objects and their children.

Reply:
Ideally yes, but not really worth it now.

Reply:
ok